File: dotnet/{{cookiecutter.project_class_name}}/{{cookiecutter.project_class_name}}.Api/Utils/ErrorDetector.cs
{%- if cookiecutter.cloud_service in ['Azure Function App', 'GCP Cloud Function'] %}
namespace {{cookiecutter.project_class_name}}.Api.Utils;

using System;
using System.Collections.Generic;
using Microsoft.AspNetCore.Mvc;
{%- if cookiecutter.cloud_service == 'Azure Function App' %}
using Microsoft.Azure.Cosmos;
{%- endif %}
{%- if cookiecutter.cloud_service == 'GCP Cloud Function' %}
using Grpc.Core;
{%- endif %}

public class BaseError
{
    public required string ErrorMessage { get; set; }
}

public class HttpResponseInit : ObjectResult
{
    public HttpResponseInit(object value, int statusCode = 500)
        : base(value)
    {
        base.StatusCode = statusCode;
    }
}

public static class ErrorDetector
{
    public static HttpResponseInit DetectError<T>(T error)
    {
{%- if cookiecutter.cloud_service == 'Azure Function App' %}
        if (error is CosmosException cosmosError)
        {
            var statusCode = (int)cosmosError.StatusCode;
            return new HttpResponseInit(new BaseError() { ErrorMessage = cosmosError.Message }, statusCode);
        }
{%- endif %}
{%- if cookiecutter.cloud_service == 'GCP Cloud Function' %}
        if (error is RpcException rpcError)
        {
            var statusCode = rpcError.StatusCode switch
            {
                StatusCode.NotFound => 404,
                StatusCode.AlreadyExists => 409,
                StatusCode.InvalidArgument => 400,
                StatusCode.PermissionDenied => 403,
                StatusCode.Unauthenticated => 401,
                _ => 500,
            };
            return new HttpResponseInit(new BaseError() { ErrorMessage = rpcError.Message }, statusCode);
        }
        if (error is KeyNotFoundException keyNotFoundError)
        {
            return new HttpResponseInit(new BaseError() { ErrorMessage = keyNotFoundError.Message }, 404);
        }
{%- endif %}
        if (error is Exception baseError)
        {
            return new HttpResponseInit(new BaseError() { ErrorMessage = baseError.Message });
        }

        return new HttpResponseInit(new BaseError() { ErrorMessage = "Unknown error occurred." });
    }
}
{%- endif %}
{%- if cookiecutter.cloud_service == 'AWS Lambda' %}
namespace {{cookiecutter.project_class_name}}.Api.Utils;

using System;
using System.Collections.Generic;
using Amazon.DynamoDBv2.Model;
using Amazon.Lambda.APIGatewayEvents;
using Newtonsoft.Json;

public class BaseError
{
    public required string ErrorMessage { get; set; }
}

public static class ErrorDetector
{
    public static APIGatewayProxyResponse DetectError<T>(T error)
    {
        var headers = new Dictionary<string, string> { { "Content-Type", "application/json" } };

        if (error is ResourceNotFoundException notFoundError)
        {
            return new APIGatewayProxyResponse
            {
                StatusCode = 404,
                Headers = headers,
                Body = JsonConvert.SerializeObject(new BaseError { ErrorMessage = notFoundError.Message })
            };
        }
        if (error is KeyNotFoundException keyNotFoundError)
        {
            return new APIGatewayProxyResponse
            {
                StatusCode = 404,
                Headers = headers,
                Body = JsonConvert.SerializeObject(new BaseError { ErrorMessage = keyNotFoundError.Message })
            };
        }
        if (error is Exception baseError)
        {
            return new APIGatewayProxyResponse
            {
                StatusCode = 500,
                Headers = headers,
                Body = JsonConvert.SerializeObject(new BaseError { ErrorMessage = baseError.Message })
            };
        }

        return new APIGatewayProxyResponse
        {
            StatusCode = 500,
            Headers = headers,
            Body = JsonConvert.SerializeObject(new BaseError { ErrorMessage = "Unknown error occurred." })
        };
    }
}

public static class ResponseHelper
{
    private static readonly Dictionary<string, string> JsonHeaders = new() { { "Content-Type", "application/json" } };

    public static APIGatewayProxyResponse Ok(object body) => new()
    {
        StatusCode = 200,
        Headers = JsonHeaders,
        Body = JsonConvert.SerializeObject(body)
    };

    public static APIGatewayProxyResponse Created(object body) => new()
    {
        StatusCode = 201,
        Headers = JsonHeaders,
        Body = JsonConvert.SerializeObject(body)
    };
}
{%- endif %}

File: go/template/utils/error_detector.go
package utils

import (
	"encoding/json"
{%- if cloud_service == 'Azure Function App' or cloud_service == 'GCP Cloud Function' %}
	"net/http"
{%- endif %}
{%- if cloud_service == 'AWS Lambda' %}
	"github.com/aws/aws-lambda-go/events"
{%- endif %}

	"{{project_endpoint}}/models"
)
{% if cloud_service == 'Azure Function App' or cloud_service == 'GCP Cloud Function' %}
func DetectError(w http.ResponseWriter, err error) {
	w.Header().Set("Content-Type", "application/json")

	var statusCode int

	switch err.(type) {
	case *models.NotFoundError:
		statusCode = http.StatusNotFound
	case *models.ValidationError:
		statusCode = http.StatusBadRequest
	default:
		statusCode = http.StatusInternalServerError
	}

	w.WriteHeader(statusCode)
	json.NewEncoder(w).Encode(models.BaseError{ErrorMessage: err.Error()})
}
{%- endif %}
{%- if cloud_service == 'AWS Lambda' %}
func DetectError(err error) events.APIGatewayProxyResponse {
	var statusCode int

	switch err.(type) {
	case *models.NotFoundError:
		statusCode = 404
	case *models.ValidationError:
		statusCode = 400
	default:
		statusCode = 500
	}

	return GenerateErrorResponse(err.Error(), statusCode)
}

func GenerateErrorResponse(message string, statusCode int) events.APIGatewayProxyResponse {
	body, _ := json.Marshal(models.BaseError{ErrorMessage: message})
	return events.APIGatewayProxyResponse{
		StatusCode: statusCode,
		Headers:    map[string]string{"Content-Type": "application/json"},
		Body:       string(body),
	}
}
{%- endif %}
